ui/ducks/send/send.js

```diff
@@ -0,0 +1,1448 @@
+        // original 'from' account, which may differ from the selected account
+        if (state.stage !== SEND_STAGES.EDIT) {
+          // This event occurs when the user selects a new account from the
+          // account menu, or the currently active account's balance updates.
+          state.account.balance = action.payload.account.balance;
+          state.account.address = action.payload.account.address;
+          // We need to update the asset balance if the asset is the native
+          // network asset. Once we update the balance we recompute error state.
+          if (state.asset.type === ASSET_TYPES.NATIVE) {
+            state.asset.balance = action.payload.account.balance;
+          }
+          slice.caseReducers.validateAmountField(state);
+          slice.caseReducers.validateGasField(state);
+          slice.caseReducers.validateSendState(state);
+        }
+      })
+      .addCase(ACCOUNT_CHANGED, (state, action) => {
+        // If we are on the edit flow then we need to watch for changes to the
+        // current account.address in state and keep balance updated
+        // appropriately
+        if (
+          state.stage === SEND_STAGES.EDIT &&
+          action.payload.account.address === state.account.address
+        ) {
+          // This event occurs when the user's account details update due to
+          // background state changes. If the account that is being updated is
+          // the current from account on the edit flow we need to update
+          // the balance for the account and revalidate the send state.
+          state.account.balance = action.payload.account.balance;
+          // We need to update the asset balance if the asset is the native
+          // network asset. Once we update the balance we recompute error state.
+          if (state.asset.type === ASSET_TYPES.NATIVE) {
+            state.asset.balance = action.payload.account.balance;
+          }
+          slice.caseReducers.validateAmountField(state);
+          slice.caseReducers.validateGasField(state);
+          slice.caseReducers.validateSendState(state);
+        }
+      })
+      .addCase(ADDRESS_BOOK_UPDATED, (state, action) => {
+        // When the address book updates from background state changes we need
+        // to check to see if an entry exists for the current address or if the
+        // entry changed.
+        const { addressBook } = action.payload;
+        if (addressBook[state.recipient.address]?.name) {
+          state.recipient.nickname = addressBook[state.recipient.address].name;
+        }
+      })
+      .addCase(initializeSendState.pending, (state) => {
+        // when we begin initializing state, which can happen when switching
+        // chains even after loading the send flow, we need to temporarily
+        // prevent the form from being submitted to allow for related state
+        // changes to be pulled in.
+        state.status = SEND_STATUSES.INVALID;
```

Comment:
Instead of using the `status` state to represent the loading state of these calls, we might want to consider storing the initialization state and the gas loading state independently, considering it in `validateSendState`, and calling it whenever these states change.

Otherwise we risk incorrectly marking the state as valid when something is still loading. For example, imagine that the send slice is being initialized at the same time as the gas estimate is loading (maybe as a result of a network switch, so there is existing send state). Then as a result of `computeEstimatedGasLimit.fulfilled`, the `status` can be set to valid despite the send initialization not being finished yet.

We need to separate these states to avoid this being confused and overloaded.

Reply:
I sort of did this here: https://github.com/MetaMask/metamask-extension/pull/10965/commits/fe30b122f4a3505288040fc6828646a87f68d6b2

When call initializeSendState in two cases: when the chain changes and when the send page is first loaded. With this change and one of our stages being UNINITIALIZED we do sort of track the initialized state of the send slice separate from gas loading. 

To take this further, though, I would need to rethink the initial stage. We can't, with the current slice/setup change stage BACK to uninitialized because that would cause the content rendered in the send page to change. I would probably need to remove UNINITIALIZED from 'stages' and make it its own state entry (isInitialized?) and keep the initial stage set as 'ADD_RECIPIENT' which is always the first thing a user should see except when editing. 

Thoughts?

Reply:
This seems reasonable to me!

Reply:
Your plans on how to take this further seem a bit cleaner, but I can't think of a compelling reason to do that right now if what we have already works. And it seems correct as far as I can tell.

Reply:
Oh! I see one part was missed though - the `computeEstimatedGasLimit.pending` event. It is still setting the status to INVALID, rather than setting this new gas loading variable to true.

Reply:
https://github.com/MetaMask/metamask-extension/pull/10965/commits/581597d686ef24e6512300abe964d212446b6f8f Good catch, i also had to rename it from isGas**Price**EstimateLoading to just isGasEstimateLoading since it should consider both sub fields to the estimate.